### ISSUE TITLE
Emscripten C struck at function socket()

### ISSUE THREAD

**Issue by AithaTarun** [emscripten-core/emscripten#17972]
```
int accept_connection (int port)
{
    int bound;
    int sock;
    struct sockaddr_in sin;
    int on = 1;
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = INADDR_ANY;
    sin.sin_port = htons (/*VNC_BASE + port*/8080);

    printf("In\n");  // <-- Getting printed
    
    bound = socket (AF_INET, SOCK_STREAM, IPPROTO_IP); // <-- Getting struck HERE

    printf("Out\n");  // <-- Not Getting printed

    if (bound < 0) {
        perror ("socket");
        exit (1);
    }

    setsockopt (bound, SOL_SOCKET, SO_REUSEADDR, &on, sizeof (on));

    if (bind (bound, (struct sockaddr *) &sin,
          sizeof (struct sockaddr_in))) {
        perror ("bind");
        exit (1);
    }
    listen (bound, 1);
    sock = accept (bound, NULL, 0);
    close (bound);
    return sock;
}
```

int accept_connection (int port)
{
    int bound;
    int sock;
    struct sockaddr_in sin;
    int on = 1;
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = INADDR_ANY;
    sin.sin_port = htons (/*VNC_BASE + port*/8080);

    printf("In\n");  // <-- Getting printed
    
    bound = socket (AF_INET, SOCK_STREAM, IPPROTO_IP); // <-- Getting struck HERE

    printf("Out\n");  // <-- Not Getting printed

    if (bound < 0) {
        perror ("socket");
        exit (1);
    }

    setsockopt (bound, SOL_SOCKET, SO_REUSEADDR, &on, sizeof (on));

    if (bind (bound, (struct sockaddr *) &sin,
          sizeof (struct sockaddr_in))) {
        perror ("bind");
        exit (1);
    }
    listen (bound, 1);
    sock = accept (bound, NULL, 0);
    close (bound);
    return sock;
}

In the above function, program execution is going into an infinite loop after calling the socket() function, I have built this with emscripten with the command :

`emcc ./rfbproxy.c -s WASM=1 -o /home/tarun/Projects/webassembly/public/rfbproxy.js -s FORCE_FILESYSTEM --embed-file tests -s "EXPORTED_RUNTIME_METHODS=['FS']" -lwebsocket.js -s PROXY_POSIX_SOCKETS=1 -s USE_PTHREADS=1 -s PROXY_TO_PTHREAD=1`

I am using Emscripten of version 3.1.23

**Comment by sbc100** [comment]
Do you follow the instruction in https://emscripten.org/docs/porting/networking.html#full-posix-sockets-over-websocket-proxy-server about setting up the websocket proxy server?

**Comment by AithaTarun** [comment]
> Do you follow the instruction in https://emscripten.org/docs/porting/networking.html#full-posix-sockets-over-websocket-proxy-server about setting up the websocket proxy server?

Yes while building .wasm I am using the flags -lwebsocket.js -s PROXY_POSIX_SOCKETS=1 -s USE_PTHREADS=1 -s PROXY_TO_PTHREAD=1

I am using [websockify](https://github.com/novnc/websockify) to connect and communicate with sockets. If I am wrong can you please explain the procedure to make this work.

**Comment by sbc100** [comment]
Are you also building the `tools/websocket_to_posix_proxy/` server and running it on the server side?  I think `-sPROXY_POSIX_SOCKETS=1` requires this.

**Comment by AithaTarun** [comment]
> Are you also building the `tools/websocket_to_posix_proxy/` server and running it on the server side? I think `-sPROXY_POSIX_SOCKETS=1` requires this.

No, by using websockify to run the server and trying to connect to it through this.

**Comment by AithaTarun** [comment]
> Are you also building the `tools/websocket_to_posix_proxy/` server and running it on the server side? I think `-sPROXY_POSIX_SOCKETS=1` requires this.

Here I think the issue is occurring while creating a socket not while communicating through the socket

**Comment by sbc100** [comment]
According the documentation if you want the default websockify approach then you should be using `-sWEBSOCKET_URL` and maybe `-sWEBSOCKET_SUBPROTOCOL`, but not `-sPROXY_POSIX_SOCKETS`.   

That is just my reading of https://emscripten.org/docs/porting/networking.html#full-posix-sockets-over-websocket-proxy-server and https://emscripten.org/docs/porting/networking.html#emulated-posix-tcp-sockets-over-websockets though.  I've never actually used this stuff outside of the testing.

**Comment by AithaTarun** [comment]
> According the documentation if you want the default websockify approach then you should be using `-sWEBSOCKET_URL` and maybe `-sWEBSOCKET_SUBPROTOCOL`, but not `-sPROXY_POSIX_SOCKETS`.
> 
> That is just my reading of https://emscripten.org/docs/porting/networking.html#full-posix-sockets-over-websocket-proxy-server and https://emscripten.org/docs/porting/networking.html#emulated-posix-tcp-sockets-over-websockets though. I've never actually used this stuff outside of the testing.

Ok thanks, I will give that a try

**Comment by AithaTarun** [comment]
> According the documentation if you want the default websockify approach then you should be using `-sWEBSOCKET_URL` and maybe `-sWEBSOCKET_SUBPROTOCOL`, but not `-sPROXY_POSIX_SOCKETS`.
> 
> That is just my reading of https://emscripten.org/docs/porting/networking.html#full-posix-sockets-over-websocket-proxy-server and https://emscripten.org/docs/porting/networking.html#emulated-posix-tcp-sockets-over-websockets though. I've never actually used this stuff outside of the testing.

I have used the below command to compile : ( and running websockify parallelly)
`emcc ./test.c -o ./test.js -sWEBSOCKET_URL="ws//localhost:8080/" -sWEBSOCKET_SUBPROTOCOL="text" -lwebsocket.js
`
Still, I am facing the same issue

**Comment by sbc100** [comment]
I'm afraid I don't have any more tips other than to start debugging.  Presumably you can see the websocket being created in the devtools, and websockify and log/trace anything that is happening on the server?

**Comment by AithaTarun** [comment]
> I'm afraid I don't have any more tips other than to start debugging. Presumably you can see the websocket being created in the devtools, and websockify and log/trace anything that is happening on the server?

Sorry, above command solved the issue (not getting struck at socket method). But now I am getting below error, and script is not waiting at accept() to start communication

![image](https://user-images.githubusercontent.com/67645139/193827780-abd526be-b25a-47d9-b6f4-394e90774aba.png)

**Comment by sbc100** [comment]
It looks like `setsockopt` is missing from the emulation when running in that mode.  The JS implementation for the socket syscalls is in src/library_socket.js.  I see an implementation of getsockopt here: https://github.com/emscripten-core/emscripten/blob/main/src/library_syscall.js#L400.  But no setsockopt. 

The setsockopt call is falling back to the stub here: https://github.com/emscripten-core/emscripten/blob/d9b9f0c2c0bf3fcb895b509ba1ca7147a7e68010/system/lib/libc/emscripten_syscall_stubs.c#L246-L249

If you would like to work on this then you could remove the stub and implement the API in JS. 

Alternatively you could look at you code and see what the setsockopt is actually doing?  Is it doing something essential or could it be removed in the emscripten build?

**Comment by sbc100** [comment]
Regarding the accept issue, I don't think we support blocking at the moment.  

The implementation for accept is here: https://github.com/emscripten-core/emscripten/blob/d9b9f0c2c0bf3fcb895b509ba1ca7147a7e68010/src/library_syscall.js#L347-L358

Then ends up calling into the filesytem code: https://github.com/emscripten-core/emscripten/blob/d9b9f0c2c0bf3fcb895b509ba1ca7147a7e68010/src/library_sockfs.js#L559-L566

I don't see any support for blocking in there.

**Comment by AithaTarun** [comment]
> 

int setsockopt(int socket, int level, int option_name,  const void *option_value, socklen_t option_len);
It sets some important options associated with a socket. Without it my C code itself not working, unable to open the connection

**Comment by AithaTarun** [comment]
> It looks like `setsockopt` is missing from the emulation when running in that mode. The JS implementation for the socket syscalls is in src/library_socket.js. I see an implementation of getsockopt here: https://github.com/emscripten-core/emscripten/blob/main/src/library_syscall.js#L400. But no setsockopt.
> 
> The setsockopt call is falling back to the stub here:
> 
> https://github.com/emscripten-core/emscripten/blob/d9b9f0c2c0bf3fcb895b509ba1ca7147a7e68010/system/lib/libc/emscripten_syscall_stubs.c#L246-L249
> 
> If you would like to work on this then you could remove the stub and implement the API in JS.
> 
> Alternatively you could look at you code and see what the setsockopt is actually doing? Is it doing something essential or could it be removed in the emscripten build?

int setsockopt(int socket, int level, int option_name, const void *option_value, socklen_t option_len);
It sets some important options associated with a socket. Without it my C code itself not working, unable to open the connection

**Comment by AithaTarun** [comment]
> Regarding the accept issue, I don't think we support blocking at the moment.
> 
> The implementation for accept is here:
> 
> https://github.com/emscripten-core/emscripten/blob/d9b9f0c2c0bf3fcb895b509ba1ca7147a7e68010/src/library_syscall.js#L347-L358
> 
> Then ends up calling into the filesytem code:
> 
> https://github.com/emscripten-core/emscripten/blob/d9b9f0c2c0bf3fcb895b509ba1ca7147a7e68010/src/library_sockfs.js#L559-L566
> 
> I don't see any support for blocking in there.

Ok, it might work if we resolve setsockopt issue

**Comment by AithaTarun** [comment]
> It looks like `setsockopt` is missing from the emulation when running in that mode. The JS implementation for the socket syscalls is in src/library_socket.js. I see an implementation of getsockopt here: https://github.com/emscripten-core/emscripten/blob/main/src/library_syscall.js#L400. But no setsockopt.
> 
> The setsockopt call is falling back to the stub here:
> 
> https://github.com/emscripten-core/emscripten/blob/d9b9f0c2c0bf3fcb895b509ba1ca7147a7e68010/system/lib/libc/emscripten_syscall_stubs.c#L246-L249
> 
> If you would like to work on this then you could remove the stub and implement the API in JS.
> 
> Alternatively you could look at you code and see what the setsockopt is actually doing? Is it doing something essential or could it be removed in the emscripten build?

So, sbc100 is there another way to achieve this functionality, or should we implement the new code?

**Comment by sbc100** [comment]
For the `setsockopt` code is really depends what options are trying to be set.  What arguments are you code passing to `setsockopt`?

**Comment by AithaTarun** [comment]
> For the `setsockopt` code is really depends what options are trying to be set. What arguments are you code passing to `setsockopt`?

Same has this one : setsockopt (bound, SOL_SOCKET, SO_REUSEADDR, &on, sizeof (on));
return value of socket, and address of int on

**Comment by sbc100** [comment]
I think it should be safe to add a version of `setsockopt` to library_syscalls.js that just accepts `SO_REUSEADDR` and does nothing.  

Alternatively you can just remove that call from you code (or wrap it in `#ifndef __EMSCRIPTEN__`)

**Comment by AithaTarun** [comment]
> I think it should be safe to add a version of `setsockopt` to library_syscalls.js that just accepts `SO_REUSEADDR` and does nothing.
> 
> Alternatively you can just remove that call from you code (or wrap it in `#ifndef __EMSCRIPTEN__`)

If I remove that call from my code, my code is not functioning, so, I think we should try adding a version of setsockopt to library_syscall.js

**Comment by sbc100** [comment]
`SO_REUSEADDR` just allows your program to re-bind to the same address next time it runs.  I doubt that is the cause of your issue.

**Comment by AithaTarun** [comment]
> `SO_REUSEADDR` just allows your program to re-bind to the same address next time it runs. I doubt that is the cause of your issue.

yes, thanks I think I can make it work. But now as we know when we execute the C code of this after creating the socket program wait at the accept method for the client to start communication, I am unable to achieve that functionality here. So, can we implement that blocking functionality, also, I don't know while running emscripten I am also getting listen status as -1 instead of 0.

**Comment by AithaTarun** [comment]
> > `SO_REUSEADDR` just allows your program to re-bind to the same address next time it runs. I doubt that is the cause of your issue.
> 
> yes, thanks I think I can make it work. But now as we know when we execute the C code of this after creating the socket; the program waits at the accept method for the client to start communication, I am unable to achieve that functionality here. So, can we implement that blocking functionality, also, I don't know while running emscripten I am also getting listen status as -1 instead of 0.

**Comment by sbc100** [comment]
Can you try linking with `-sSOCKET_DEBUG` and see what shows up in the console?

**Comment by AithaTarun** [comment]
> Can you try linking with `-sSOCKET_DEBUG` and see what shows up in the console?

Nothing new :
![image](https://user-images.githubusercontent.com/67645139/194092143-ebf0a832-6e0f-4004-ad2a-76963a96cbdb.png)

**Comment by sbc100** [comment]
Actually it like make `listen` is not supported at all outside of node:

https://github.com/emscripten-core/emscripten/blob/40fb7d2071e439f1de614898b88518df582faa94/src/library_sockfs.js#L499-L503

I think you need to use `PROXY_POSIX_SOCKETS` if you really need to receive incoming connections.   Setting that requires the building and running of the POSIX Sockets proxy server program.  Please make sure you read through all documentation at https://emscripten.org/docs/porting/networking.html.

**Comment by AithaTarun** [comment]
> 

ok thanks, once again I will give that a try

**Comment by AithaTarun** [comment]
> Actually it like make `listen` is not supported at all outside of node:
> 
> https://github.com/emscripten-core/emscripten/blob/40fb7d2071e439f1de614898b88518df582faa94/src/library_sockfs.js#L499-L503
> 
> I think you need to use `PROXY_POSIX_SOCKETS` if you really need to receive incoming connections. Setting that requires the building and running of the POSIX Sockets proxy server program. Please make sure you read through all documentation at https://emscripten.org/docs/porting/networking.html.

ok thanks, once again I will give that a try

**Comment by nikitapn** [comment]
To use `Full POSIX Sockets over WebSocket Proxy Server` you need to call 
> #include <emscripten/posix_socket.h> emscripten_init_websocket_to_posix_socket_bridge("ws://bridge_server_ip_port")

Also need some delay after call to this function, before any socket() invocation.

**Comment by R2Sam** [comment]
I have a very similar issue using enet. Through some log spamming it seems it is stalling at the socket() call within enet. Using these for linking `-lwebsocket.js -sPROXY_POSIX_SOCKETS -pthread -sPROXY_TO_PTHREAD`
